master/internal/rm/kubernetesrm/pods.go

```diff
@@ -934,99 +948,103 @@ func (p *pods) handleGetAgentsRequest(ctx *actor.Context) {
 	ctx.Respond(response)
 }
 
-// summarize describes pods' available resources. When there's exactly one resource pool and that
-// pool has no quotas configured, it uses the whole cluster's info. Otherwise, it uses namespaces'
-// quotas to derive that info.
+// summarize describes pods' available resources. When there's exactly one resource pool, it uses
+// the whole cluster's info. Otherwise, it matches nodes to resource pools using taints and
+// tolerations to derive that info.
 func (p *pods) summarize(ctx *actor.Context) (map[string]model.AgentSummary, error) {
-	namespaceToQuota := make(map[string]k8sV1.ResourceQuota)
+	nodeSummaries := p.summarizeClusterByNodes(ctx)
 
-	// Look up quotas for our resource pools' namespaces.
-	for namespace := range p.namespaceToPoolName {
-		quotaList, err := p.quotaInterfaces[namespace].List(context.TODO(), metaV1.ListOptions{})
-		if err != nil && !k8serrors.IsNotFound(err) {
-			return nil, err
-		} else if k8serrors.IsNotFound(err) || quotaList == nil {
-			continue
-		}
+	taskContainerDefaults := extractTCDs(p.resourcePoolConfigs)
 
-		relevantQuotas := cpuAndGpuQuotas(quotaList)
-		if len(relevantQuotas) != 1 {
-			// TODO: figure out how we want to handle multiple quotas per namespace?
-			// When there's multiple conflicting quotas, k8s seems to use the most
-			// restrictive of them—i.e. if there's a quota limiting to 100 CPUs and one
-			// limiting to 10, only 10 CPUs will be allowed.
-			continue
-		}
+	// Build the many-to-many relationship between nodes and resource pools
+	poolsToNodes := make(map[string][]*k8sV1.Node, len(p.namespaceToPoolName))
+	for _, node := range p.currentNodes {
+		_, slotType := extractSlotInfo(nodeSummaries[node.Name])
+
+		for poolName, tcd := range taskContainerDefaults {
+			var poolTolerations []k8sV1.Toleration
+
+			if tcd != nil {
+				// Decide which poolTolerations to use based on slot device type
+				if slotType == device.CUDA && tcd.GPUPodSpec != nil {
+					poolTolerations = tcd.GPUPodSpec.Spec.Tolerations
+				} else if tcd.CPUPodSpec != nil {
+					poolTolerations = tcd.CPUPodSpec.Spec.Tolerations
+				} else {
+					logrus.WithFields(logrus.Fields{
+						"poolName": poolName,
+						"nodeName": node.Name,
+					}).Debug("didn't find appropriate podspec")
+				}
+			} else {
+				logrus.WithFields(logrus.Fields{
+					"poolName": poolName,
```

Comment:
question: If the tasks don't exist inside of a resource pool, where do they exist?

Reply:
Nodes belong in a pool, and tasks assigned to be executed on a resource pool should execute on one of that pool's nodes. A node that doesn't belong to a resource pool is obviously a bad configuration, but that should be beyond the scope of this endpoint to remedy if that's what you mean.

Or do you mean what happens when somebody tries to execute a task and doesn't restrict it to a resource pool? If the latter, I don't know!

Reply:
In case this is the source of your question: this log statement would be run in the case that a resource pool had no configured task container defaults. That would mean it had no Tolerations, so `poolTolerations` staying `nil` is desired behavior.

Note: I plan to remove the log statements I added once I'm done testing this in GKE with GPUs.

Reply:
After testing a bit, I understand more and it's tracks with what you're saying.

there always exists at least 1 resource pool and whether or not a node belongs to them depends on tolerations and whether task container defaults is set or not.